tms/models/tms_order.py

```diff
@@ -0,0 +1,261 @@
+# Copyright (C) 2024 Open Source Integrators
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from datetime import datetime, timedelta
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class TMSOrder(models.Model):
+    _name = "tms.order"
+    _description = "Transport Management System Order"
+    _inherit = ["mail.thread", "mail.activity.mixin"]
+
+    active = fields.Boolean(default=True)
+
+    name = fields.Char(
+        required=True,
+        copy=False,
+        readonly=False,
+        index="trigram",
+        default=lambda self: _("New"),
+    )
+
+    company_id = fields.Many2one(
+        "res.company",
+        string="Company",
+        required=True,
+        index=True,
+        default=lambda self: self.env.company,
+        help="Company related to this order",
+    )
+
+    description = fields.Text()
+    sequence = fields.Integer(default=10)
+
+    route = fields.Boolean(string="Use predefined route")
+    route_id = fields.Many2one("tms.route")
+    route_origin = fields.Char(related="route_id.origin_location_id.display_name")
+    route_destination = fields.Char(
+        related="route_id.destination_location_id.display_name"
+    )
+
+    origin_id = fields.Many2one(
+        "res.partner",
+        domain="[('tms_type', '=', 'location')]",
+        context={"default_tms_type": "location"},
+    )
+    destination_id = fields.Many2one(
+        "res.partner",
+        domain="[('tms_type', '=', 'location')]",
+        context={"default_tms_type": "location"},
+    )
+
+    origin_location = fields.Char(string="Origin")
+    destination_location = fields.Char(string="Destination")
+
+    driver_id = fields.Many2one(
+        "res.partner",
+        string="Driver",
+        context={"default_tms_type": "driver"},
+    )
+    vehicle_id = fields.Many2one("fleet.vehicle", string="Vehicle")
+    team_id = fields.Many2one("tms.team", string="Team")
+    crew_id = fields.Many2one("tms.crew", string="Crew")
+
+    stage_id = fields.Many2one(
+        "tms.stage",
+        string="Stage",
+        index=True,
+        copy=False,
+        domain="[('stage_type', '=', 'order')]",
+        default=lambda self: self._default_stage_id(),
+        group_expand="_read_group_stage_ids",
+        ondelete="set null",
+    )
+    stage = fields.Char(related="stage_id.name")
+
+    scheduled_date_start = fields.Datetime(
+        string="Scheduled Start", default=datetime.now()
+    )
+    scheduled_duration = fields.Float(
+        help="Scheduled duration of the work in" " hours",
+    )
+    scheduled_date_end = fields.Datetime(string="Scheduled End")
+
+    start_trip = fields.Boolean(readonly=True)
+    end_trip = fields.Boolean(readonly=True)
+    date_start = fields.Datetime(readonly=True)
+    date_end = fields.Datetime(readonly=True)
+    duration = fields.Float(readonly=True, string="Trip Duration")
```

Comment:
how would these fields be handled in case maintenance is needed?

Reply:
@marcelsavegnago If maintenance of the vehicle is needed, you would not be able to schedule a trip with that vehicle...

Reply:
I'm referring to data maintenance. From what I understand, the start and end of the trip are recorded when the 'Start' and 'End' buttons are clicked. Therefore, if someone doesn't click 'Start' or 'Finish', we would need to change the recorded values.